All but first test in a class fail if class is annotated @WithFunction

### Describe the bug

Given a project which contains an implementation of a Google Cloud Run Function (`CloudEventsFunction` variant) which was created as shown in the [Google Cloud Functions (Serverless) guide](https://quarkus.io/guides/gcp-functions#creating-the-maven-deployment-project),

Add a test class to this project as shown in the [Testing your function section](https://quarkus.io/guides/gcp-functions#the-cloudeventsfunction-4) and add a second identical test is added to this test class,

When running tests, the second test in this class cannot be instantiated.


### Expected behavior

Both tests pass.

### Actual behavior

The first test passes, the second test fails.

```
org.junit.jupiter.api.extension.TestInstantiationException: Failed to create test instance
        at io.quarkus.test.junit.QuarkusTestExtension.initTestState(QuarkusTestExtension.java:773)
        at io.quarkus.test.junit.QuarkusTestExtension.interceptTestClassConstructor(QuarkusTestExtension.java:739)
        at java.base/java.util.Optional.orElseGet(Optional.java:364)
        at java.base/java.util.ArrayList.forEach(ArrayList.java:1596)
        at java.base/java.util.ArrayList.forEach(ArrayList.java:1596)
Caused by: java.lang.RuntimeException: java.io.IOException: Failed to bind to 0.0.0.0/0.0.0.0:8081
        at io.quarkus.google.cloud.functions.test.CloudFunctionTestResource.inject(CloudFunctionTestResource.java:40)
        at io.quarkus.test.common.TestResourceManager.inject(TestResourceManager.java:201)
        at java.base/java.lang.reflect.Method.invoke(Method.java:580)
        at io.quarkus.test.junit.QuarkusTestExtension.createActualTestInstance(QuarkusTestExtension.java:784)
        at io.quarkus.test.junit.QuarkusTestExtension.initTestState(QuarkusTestExtension.java:767)
        ... 4 more
Caused by: java.io.IOException: Failed to bind to 0.0.0.0/0.0.0.0:8081
        at org.eclipse.jetty.server.ServerConnector.openAcceptChannel(ServerConnector.java:349)
        at org.eclipse.jetty.server.ServerConnector.open(ServerConnector.java:310)
        at org.eclipse.jetty.server.AbstractNetworkConnector.doStart(AbstractNetworkConnector.java:80)
        at org.eclipse.jetty.server.ServerConnector.doStart(ServerConnector.java:234)
        at org.eclipse.jetty.util.component.AbstractLifeCycle.start(AbstractLifeCycle.java:73)
        at org.eclipse.jetty.server.Server.doStart(Server.java:401)
        at org.eclipse.jetty.util.component.AbstractLifeCycle.start(AbstractLifeCycle.java:73)
        at com.google.cloud.functions.invoker.runner.Invoker.startServer(Invoker.java:324)
        at com.google.cloud.functions.invoker.runner.Invoker.startTestServer(Invoker.java:276)
        at io.quarkus.google.cloud.functions.test.CloudFunctionsInvoker.start(CloudFunctionsInvoker.java:26)
        at io.quarkus.google.cloud.functions.test.CloudFunctionTestResource.inject(CloudFunctionTestResource.java:38)
        ... 8 more
Caused by: java.net.BindException: Address already in use: bind
        at java.base/sun.nio.ch.Net.bind0(Native Method)
        at java.base/sun.nio.ch.Net.bind(Net.java:565)
        at java.base/sun.nio.ch.ServerSocketChannelImpl.netBind(ServerSocketChannelImpl.java:344)
        at java.base/sun.nio.ch.ServerSocketChannelImpl.bind(ServerSocketChannelImpl.java:301)
        at java.base/sun.nio.ch.ServerSocketAdaptor.bind(ServerSocketAdaptor.java:89)
        at org.eclipse.jetty.server.ServerConnector.openAcceptChannel(ServerConnector.java:344)
        ... 18 more
```


### How to Reproduce?

https://github.com/devopsix/quarkus-google-cloud-functions-issue

### Output of `uname -a` or `ver`

n/a

### Output of `java -version`

openjdk version "21.0.5" 2024-10-15 LTS OpenJDK Runtime Environment Temurin-21.0.5+11 (build 21.0.5+11-LTS) OpenJDK 64-Bit Server VM Temurin-21.0.5+11 (build 21.0.5+11-LTS, mixed mode, sharing)

### Quarkus version or git rev

3.16.4

### Build tool (ie. output of `mvnw --version` or `gradlew --version`)

Apache Maven 3.9.9 (8e8579a9e76f7d015ee5ec7bfcdc97d260186937)

### Additional information

Suspected cause: In [`CloudFunctionTestResource.inject(TestInjector)`](https://github.com/quarkusio/quarkus/blob/8df922d7f8a07ef97a73ba066f2496486140eb67/test-framework/google-cloud-functions/src/main/java/io/quarkus/google/cloud/functions/test/CloudFunctionTestResource.java#L32) a new invoker is started for each test without stopping an invoker created previously.

Potential remedy: Invokers should either be reused between tests or stopped after each test.

--- a/integration-tests/google-cloud-functions/src/test/java/io/quarkus/gcp/function/test/HttpFunctionTestCase.java
+++ b/integration-tests/google-cloud-functions/src/test/java/io/quarkus/gcp/function/test/HttpFunctionTestCase.java
@@ -21,4 +21,15 @@ class HttpFunctionTestCase {
                 .statusCode(200)
                 .body(is("Hello World!"));
     }
+
+    // we do twice the test to be sure we can call multiple time the same function
+    @Test
+    public void test2() {
+        // test the function using RestAssured
+        when()
+                .get()
+                .then()
+                .statusCode(200)
+                .body(is("Hello World!"));
+    }
 }

--- a/test-framework/google-cloud-functions/src/main/java/io/quarkus/google/cloud/functions/test/CloudFunctionTestResource.java
+++ b/test-framework/google-cloud-functions/src/main/java/io/quarkus/google/cloud/functions/test/CloudFunctionTestResource.java
@@ -2,6 +2,7 @@ package io.quarkus.google.cloud.functions.test;
 
 import java.util.Collections;
 import java.util.Map;
+import java.util.concurrent.atomic.AtomicBoolean;
 
 import org.eclipse.microprofile.config.ConfigProvider;
 
@@ -16,6 +17,7 @@ public class CloudFunctionTestResource implements QuarkusTestResourceConfigurabl
     private FunctionType functionType;
     private String functionName;
     private CloudFunctionsInvoker invoker;
+    private final AtomicBoolean started = new AtomicBoolean(false);
 
     @Override
     public void init(WithFunction withFunction) {
@@ -30,14 +32,16 @@ public class CloudFunctionTestResource implements QuarkusTestResourceConfigurabl
 
     @Override
     public void inject(TestInjector testInjector) {
-        // This is a hack, we cannot start the invoker in the start() method as Quarkus is not yet initialized,
-        // so we start it here as this method is called later (the same for reading the test port).
-        int port = ConfigProvider.getConfig().getOptionalValue("quarkus.http.test-port", Integer.class).orElse(8081);
-        this.invoker = new CloudFunctionsInvoker(functionType, port);
-        try {
-            this.invoker.start();
-        } catch (Exception e) {
-            throw new RuntimeException(e);
+        if (started.compareAndSet(false, true)) {
+            // This is a hack, we cannot start the invoker in the start() method as Quarkus is not yet initialized,
+            // so we start it here as this method is called later (the same for reading the test port).
+            int port = ConfigProvider.getConfig().getOptionalValue("quarkus.http.test-port", Integer.class).orElse(8081);
+            this.invoker = new CloudFunctionsInvoker(functionType, port);
+            try {
+                this.invoker.start();
+            } catch (Exception e) {
+                throw new RuntimeException(e);
+            }
         }
     }
 

